server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java

```diff
@@ -146,28 +148,28 @@ private void rangeCheck(Translog.Operation op) {
         }
     }
 
-    private int nextDocId() {
-        if (docIndex < topDocs.scoreDocs.length) {
-            final int docId = topDocs.scoreDocs[docIndex].doc;
+    private int nextDocId() throws IOException {
+        // we have processed all docs in the current search - fetch the next window
+        if (docIndex == scoreDocs.length && docIndex > 0) {
+            final ScoreDoc prev = scoreDocs[scoreDocs.length - 1];
+            scoreDocs = searchOperations(prev).scoreDocs;
+            docIndex = 0;
+        }
+        if (docIndex < scoreDocs.length) {
+            int docId = scoreDocs[docIndex].doc;
             docIndex++;
             return docId;
-        } else {
-            return DocIdSetIterator.NO_MORE_DOCS;
         }
+        return DocIdSetIterator.NO_MORE_DOCS;
     }
 
-    private TopDocs searchOperations(IndexSearcher searcher) throws IOException {
-        final Query rangeQuery = new BooleanQuery.Builder()
-            .add(new DocValuesFieldExistsQuery(SeqNoFieldMapper.PRIMARY_TERM_NAME), BooleanClause.Occur.FILTER)
-            .add(LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, fromSeqNo, toSeqNo), BooleanClause.Occur.FILTER)
-            .build();
+    private TopDocs searchOperations(ScoreDoc after) throws IOException {
+        final Query rangeQuery = LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, fromSeqNo, toSeqNo);
```

Comment:
why did we drop the  `DocValuesFieldExistsQuery(SeqNoFieldMapper.PRIMARY_TERM_NAME)` part?

Reply:
`DocValuesFieldExistsQuery(SeqNoFieldMapper.PRIMARY_TERM_NAME)` was in the temporary fix (https://github.com/elastic/elasticsearch/commit/7004d2c7bd070f49872da6152e05d784fe16b69c). This clause was to use to eliminate nested docs.